python/lsst/meas/astrom/astrometry.py

```diff
@@ -251,71 +251,92 @@ def solve(self, exposure, sourceCat):
-                (tryMatchDist.distMean.asArcseconds(), self.config.maxMeanDistanceArcsec))
-        for m in res.matches:
-            if self.usedKey:
-                m.second.set(self.usedKey, True)
-        exposure.setWcs(res.wcs)
+                    "Match and fit WCS iteration %d: found %d matches with on-sky distance mean and "
+                    "scatter = %0.3f +- %0.3f arcsec; max match distance = %0.3f arcsec",
+                    iterNum, len(tryRes.matches), tryMatchDist.distMean.asArcseconds(),
+                    tryMatchDist.distStdDev.asArcseconds(), tryMatchDist.maxMatchDist.asArcseconds())
+
+                maxMatchDist = tryMatchDist.maxMatchDist
+                res = tryRes
+                wcs = res.wcs
+                if maxMatchDist.asArcseconds() < self.config.minMatchDistanceArcSec:
+                    self.log.debug(
+                        "Max match distance = %0.3f arcsec < %0.3f = config.minMatchDistanceArcSec; "
+                        "that's good enough",
+                        maxMatchDist.asArcseconds(), self.config.minMatchDistanceArcSec)
+                    break
+                match_tolerance.maxMatchDist = maxMatchDist
+
+        if not fitFailed:
+            self.log.info("Matched and fit WCS in %d iterations; "
+                          "found %d matches with mean and scatter = %0.3f +- %0.3f arcsec" %
+                          (iterNum, len(tryRes.matches), tryMatchDist.distMean.asArcseconds(),
+                           tryMatchDist.distStdDev.asArcseconds()))
+            if tryMatchDist.distMean.asArcseconds() > self.config.maxMeanDistanceArcsec:
+                self.log.info("Assigning as a fit failure: mean on-sky distance = %0.3f arcsec > %0.3f "
+                              "(maxMeanDistanceArcsec)" % (tryMatchDist.distMean.asArcseconds(),
+                                                           self.config.maxMeanDistanceArcsec))
+                fitFailed = True
+
+        if fitFailed:
+            self.log.warning("WCS fit failed.  Setting exposure's WCS to None and coord_ra & coord_dec "
+                             "cols in sourceCat to nan.")
+            sourceCat["coord_ra"] = np.nan
+            sourceCat["coord_dec"] = np.nan
+            exposure.setWcs(None)
+            matches = None
+            scatterOnSky = None
+        else:
+            for m in res.matches:
+                if self.usedKey:
+                    m.second.set(self.usedKey, True)
+            exposure.setWcs(res.wcs)
+            matches = res.matches
+            scatterOnSky = res.scatterOnSky
 
         # Record the scatter in the exposure metadata
         md = exposure.getMetadata()
-        md['SFM_ASTROM_OFFSET_MEAN'] = tryMatchDist.distMean.asArcseconds()
-        md['SFM_ASTROM_OFFSET_STD'] = tryMatchDist.distStdDev.asArcseconds()
+        if fitFailed:
+            md['SFM_ASTROM_OFFSET_MEAN'] = np.nan
+            md['SFM_ASTROM_OFFSET_STD'] = np.nan
```

Comment:
I'm actually uncertain about this one.  We do have a measurement of the OFFSET_MEAN and OFFSET_STD, it's just that they failed our threshold, right?  I think having these data recorded (how bad was it?) is valuable.